diambra/arena/env_settings.py

```diff
@@ -1,200 +1,389 @@
+    assert (value in valid_list), error_message
+    assert (type(value)==type(valid_list[valid_list.index(value)])), error_message
 
 @dataclass
 class EnvironmentSettings:
+    env_info = None
+    games_dict = None
 
+    # Fixed at startup
     game_id: str
-
-    # System level
+    frame_shape: Tuple[int, int, int] = (0, 0, 0)
     step_ratio: int = 6
-    disable_keyboard:bool = True
+    n_players: int = 1
+    disable_keyboard: bool = True
     disable_joystick: bool = True
+    render_mode: Union[str, None] = None
     rank: int = 0
-    seed: int = -1
     env_address: str = "localhost:50051"
     grpc_timeout: int = 600
 
-    # Game level
-    player: str = "Random"
+    # Variable at reset
+    seed: Union[int, None] = None
+    difficulty: Union[int, str] = "Random"
     continue_game: float = 0.0
-    show_final: bool = True
-    difficulty: int = 3
-    frame_shape: Tuple[int, int, int] = (0, 0, 0)
-
+    show_final: bool = False
     tower: int = 3  # UMK3 Specific
 
-    # Environment level
-    hardcore: bool = False
+    # Bookeeping variables
+    _last_seed: int = None
+    pb_model: model = None
 
-    def sanity_check(self):
-        self.games_dict = available_games(False)
+    variable_env_settings = ["seed", "difficulty", "continue_game", "show_final", "tower", "role",
+                             "characters", "outfits", "super_art", "fighting_style", "ultimate_style"]
 
-        # TODO: consider using typing.Literal to specify lists of admissible values: NOTE: It requires Python 3.8+
-        check_num_in_range("step_ratio", self.step_ratio, [1, 6])
-        check_num_in_range("rank", self.rank, [0, MAX_VAL])
-        check_num_in_range("seed", self.seed, [-1, MAX_VAL])
-        check_num_in_range("grpc_timeout", self.grpc_timeout, [0, 3600])
+    # Transforming env settings dict to pb request
+    def get_pb_request(self, init=False):
+        frame_shape = {
+            "h": self.frame_shape[0],
+            "w": self.frame_shape[1],
+            "c": self.frame_shape[2]
+        }
 
-        check_val_in_list("game_id", self.game_id, self.games_dict.keys())
-        check_val_in_list("player", self.player, ["P1", "P2", "Random", "P1P2"])
-        check_num_in_range("continue_game", self.continue_game, [MIN_VAL, 1.0])
-        check_num_in_range("difficulty", self.difficulty, self.games_dict[self.game_id]["difficulty"][:2])
+        if self.seed == None:
+            self.seed = int(time.time())
+
+        if self._last_seed != self.seed:
+            random.seed(self.seed)
+            np.random.seed(self.seed)
+            self._last_seed = self.seed
+
+        action_spaces = self._get_action_spaces()
+
+        if init is False:
+            self._process_random_values()
+
+            player_env_settings = self._get_player_specific_values()
 
+            variable_env_settings = model.EnvSettings.VariableEnvSettings(
```

Comment:
Can we come up with a better name? What makes them 'variable' are player_env_settings not 'variable' Isn't this more a 'game_env_settings' or well just 'env_setttings' vs 'player_env_settings'?

Reply:
This is partially linked to what is written above, and strongly linked with the gymnasium modification, that is very handy. The idea is the following:
Environment settings can now be grouped in 2 families:
1) Those that are specified when you instantiate the env and that cannot be changed 
2) Those that can be optionally changed at every environment reset. 

The second group is contained in the first, as you can also specify them only once at startup and never change them (as we were doing before), but it is good to have it separate.

Among the variable environment settings, there are also the player related ones. In principle, they can just be mixed with the generic variable env settings as they are all variable at reset, but then the single player VS multiplayer distinction comes into place, and it is very useful to keep them separated as it is very easy to keep each player settings separated.

Is it clear. What do you think?

Reply:
Clear that these should be separated, it's just the name. Is 'variable settings' an established term? Then let's stick with it.
If not, maybe episode_settings would make more sense? But go with what you feel is right, just wanted to point this out

Reply:
I really like the "Episode Settings" name! Thanks @discordianfish I will modify it accordingly!

Reply:
Done in this: https://github.com/diambra/arena/commit/0770bff1e963cc9baecf06833a4555388c35557c